src/track/serato/beatgrid.h

```diff
@@ -0,0 +1,185 @@
+#pragma once
+
+#include <QByteArray>
+#include <QColor>
+#include <QDataStream>
+#include <QList>
+#include <memory>
+
+#include "track/cueinfo.h"
+#include "track/taglib/trackmetadata_file.h"
+#include "util/assert.h"
+#include "util/types.h"
+
+namespace mixxx {
+
+class SeratoBeatGridNonTerminalMarker;
+class SeratoBeatGridTerminalMarker;
+
+typedef std::shared_ptr<SeratoBeatGridNonTerminalMarker> SeratoBeatGridNonTerminalMarkerPointer;
+typedef std::shared_ptr<SeratoBeatGridTerminalMarker> SeratoBeatGridTerminalMarkerPointer;
+
+class SeratoBeatGridNonTerminalMarker {
+  public:
+    SeratoBeatGridNonTerminalMarker(float positionSecs, quint32 beatsTillNextMarker)
+            : m_positionSecs(positionSecs),
+              m_beatsTillNextMarker(beatsTillNextMarker) {
+    }
+    ~SeratoBeatGridNonTerminalMarker() = default;
+
+    QByteArray dumpID3() const;
+    static SeratoBeatGridNonTerminalMarkerPointer parseID3(const QByteArray&);
+
+    float positionSecs() const {
+        return m_positionSecs;
+    }
+
+    float beatsTillNextMarker() const {
+        return m_beatsTillNextMarker;
+    }
+
+  private:
+    float m_positionSecs;
+    quint32 m_beatsTillNextMarker;
+};
+
+inline QDebug operator<<(QDebug dbg, const SeratoBeatGridNonTerminalMarker& arg) {
+    return dbg << "SeratoBeatGridNonTerminalMarker"
+               << "PositionMillis =" << arg.positionSecs()
+               << "BeatTillNextMarker = " << arg.beatsTillNextMarker();
+}
+
+class SeratoBeatGridTerminalMarker {
+  public:
+    SeratoBeatGridTerminalMarker(float positionSecs, float bpm)
+            : m_positionSecs(positionSecs),
+              m_bpm(bpm) {
+    }
+    ~SeratoBeatGridTerminalMarker() = default;
+
+    QByteArray dumpID3() const;
+    static mixxx::SeratoBeatGridTerminalMarkerPointer parseID3(const QByteArray&);
+
+    float positionSecs() const {
+        return m_positionSecs;
+    }
+
+    float bpm() const {
+        return m_bpm;
+    }
+
+  private:
+    float m_positionSecs;
+    float m_bpm;
+};
+
+inline QDebug operator<<(QDebug dbg, const SeratoBeatGridTerminalMarker& arg) {
+    return dbg << "SeratoBeatGridTerminalMarker"
+               << "PositionMillis =" << arg.positionSecs()
+               << "BPM =" << arg.bpm();
+}
+
+/// DTO for storing information from the SeratoBeatGrid tags used by the Serato
+/// DJ Pro software.
+///
+/// This class includes functions for formatting and parsing SeratoBeatGrid
+/// metadata according to the specification:
+/// https://github.com/Holzhaus/serato-tags/blob/master/docs/serato_beatgrid.md
+class SeratoBeatGrid final {
+  public:
+    SeratoBeatGrid() = default;
+    explicit SeratoBeatGrid(
```

Comment:
Doesn't need to be explicit. Inconsistent C++ weirdness. Doesn't matter but I always stumble upon this when scanning the number of arguments  ;)